bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs

```diff
@@ -0,0 +1,763 @@
+                )
+            ),
+        CreationPolicy::Never if !cluster_exists =>
+            Err(
+                ProviderError::new_with_context(
+                    Resources::Clear, format!(
+                        "The cluster '{}' does not exist and creation policy '{:?}' requires that it exist",
+                        cluster_info.name,
+                        creation_policy
+                    )
+                )
+            ),
+        CreationPolicy::Create  =>{
+            Ok((true, format!("Creation policy is '{:?}' and cluster '{}' does not exist: creating cluster", creation_policy, cluster_info.name)))
+        },
+        CreationPolicy::IfNotExists if !cluster_exists => {
+            Ok((true, format!("Creation policy is '{:?}' and cluster '{}' does not exist: creating cluster", creation_policy, cluster_info.name)))
+        },
+        CreationPolicy::IfNotExists |
+        CreationPolicy::Never => {
+            Ok((false, format!("Creation policy is '{:?}' and cluster '{}' exists: not creating cluster", creation_policy, cluster_info.name)))
+        },
+    }
+}
+
+async fn does_cluster_exist(cluster_info: &VSphereK8sClusterInfo) -> ProviderResult<bool> {
+    if let Some(kubeconfig_base64) = &cluster_info.kubeconfig_base64 {
+        base64_decode_write_file(kubeconfig_base64, TEST_CLUSTER_KUBECONFIG_PATH)
+            .await
+            .context(
+                Resources::Clear,
+                "Failed to write out kubeconfig for vSphere K8s cluster",
+            )?;
+        let kubeconfig = Kubeconfig::read_from(TEST_CLUSTER_KUBECONFIG_PATH)
+            .context(Resources::Clear, "Unable to read kubeconfig")?;
+        let config =
+            Config::from_custom_kubeconfig(kubeconfig.to_owned(), &KubeConfigOptions::default())
+                .await
+                .context(Resources::Clear, "Unable load kubeconfig")?;
+        let k8s_client = kube::client::Client::try_from(config)
+            .context(Resources::Clear, "Unable create K8s client from kubeconfig")?;
+        let k8s_nodes: Api<Node> = Api::all(k8s_client);
+        let node_list = k8s_nodes.list(&ListParams::default()).await;
+        if node_list.is_ok() {
+            // If we can query nodes from the cluster, then we've validate that the kubeconfig is valid
+            return Ok(true);
+        }
+    }
+    Ok(false)
+}
+
+async fn create_vsphere_k8s_cluster(
+    config: &VSphereK8sClusterConfig,
+    mgmt_kubeconfig_path: &str,
+    resources: &mut Resources,
+    memo: &mut ProductionMemo,
+) -> ProviderResult<()> {
+    let (metadata_url, targets_url) = tuf_repo_urls(&config.tuf_repo, resources)?;
+
+    // Set up environment variables for govc cli
+    set_govc_env_vars(config);
+
+    // Retrieve the OVA file
+    let ova_name = config.ova_name.to_owned();
+    info!("Downloading OVA '{}'", &config.ova_name);
+    let outdir = Path::new("/local/");
+    tokio::task::spawn_blocking(move || -> ProviderResult<()> {
```

Comment:
are you doing this because download_target is a blocking function? Which it is because `tough` is not async? https://github.com/awslabs/tough/issues/213

Does `spawn_blocking` created a dedicated thread? It would be safest if it does because of the multiple tokio runtimes issue that reqwest::blocking can cause.

Reply:
> Does spawn_blocking created a dedicated thread?
Yes, it creates a thread where blocking is ok.